[dynamo][fallback] Fallback when run_node fails with fake tensors

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107213
* #107179

Little cumbersome to write a bad meta implementation for testing this. The model that I am working on falls back with this error message

Example 1
~~~
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG] Traceback (most recent call last):
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1320, in get_fake_value
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return wrap_fake_exception(
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 898, in wrap_fake_exception
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return fn()
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1321, in <lambda>
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     lambda: run_node(tx.output, node, args, kwargs, nnmodule)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1392, in run_node
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     raise RuntimeError(fn_str + str(e)).with_traceback(e.__traceback__) from e
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1379, in run_node
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return node.target(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 681, in __call__
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/utils/_stats.py", line 20, in wrapper
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return fn(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_subclasses/fake_tensor.py", line 1233, in __torch_dispatch__
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self.dispatch(func, types, args, kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_subclasses/fake_tensor.py", line 1523, in dispatch
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     r = func(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 435, in __call__
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG] RuntimeError: Failed running call_function fbgemm.dense_to_jagged(*(FakeTensor(..., device='cuda:0', size=(10, 1023, 1), dtype=torch.int64), [FakeTensor(..., device='cuda:0', size=(11,), dtype=torch.int64)]), **{}):
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG] total_L is required for meta backend
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] Graph break: Unable to propagate fake values through the operator. One possible reason could be bad meta kernel implementation. from user code at:
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/hammer/modules/features/postproc.py", line 554, in <resume in forward>
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     ) = self._action_based_sampling_impl(jt_dict)
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/nn/modules/module.py", line 1527, in _call_impl
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     return forward_call(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/hammer/modules/features/postproc.py", line 298, in forward
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     removal_mask_union_jagged = torch.ops.fbgemm.dense_to_jagged(
~~~


Example 2
~~~
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] ========== TorchDynamo Stack Trace ==========
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Traceback (most recent call last):
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1379, in run_node
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return node.target(*args, **kwargs)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 681, in __call__
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] RuntimeError: fbgemm::jagged_1d_to_dense() Expected a value of type 'int' for argument 'max_sequence_length' but instead found type 'FakeTensor'.
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Position: 2
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Value: FakeTensor(..., device='cuda:0', size=(), dtype=torch.int64)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Declaration: fbgemm::jagged_1d_to_dense(Tensor values, Tensor offsets, int max_sequence_length, int padding_value) -> Tensor
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Cast error details: Unable to cast Python instance to C++ type (compile in debug mode for details)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] The above exception was the direct cause of the following exception:
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Traceback (most recent call last):
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1320, in get_fake_value
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return wrap_fake_exception(
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 898, in wrap_fake_exception
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return fn()
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1321, in <lambda>
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     lambda: run_node(tx.output, node, args, kwargs, nnmodule)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1392, in run_node
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     raise RuntimeError(fn_str + str(e)).with_traceback(e.__traceback__) from e
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1379, in run_node
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return node.target(*args, **kwargs)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 681, in __call__
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] RuntimeError: Failed running call_function fbgemm.jagged_1d_to_dense(*(), **{'values': FakeTensor(..., device='cuda:0', size=(5110,), dtype=torch.int64), 'offsets': FakeTensor(..., device='cuda:0', size=(11,), dtype=torch.int64), 'max_sequence_length': FakeTensor(..., device='cuda:0', size=(), dtype=torch.int64), 'padding_value': 0}):
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] fbgemm::jagged_1d_to_dense() Expected a value of type 'int' for argument 'max_sequence_length' but instead found type 'FakeTensor'.
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Position: 2
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Value: FakeTensor(..., device='cuda:0', size=(), dtype=torch.int64)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Declaration: fbgemm::jagged_1d_to_dense(Tensor values, Tensor offsets, int max_sequence_length, int padding_value) -> Tensor
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Cast error details: Unable to cast Python instance to C++ type (compile in debug mode for details)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] Graph break: Unable to propagate fake values through the operator. One possible reason could be bad meta kernel implementation. from user code at:
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/hammer/modules/features/postproc.py", line 141, in <resume in flip_batched_id_list_feature>
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     dense_values = torch.ops.fbgemm.jagged_1d_to_dense(
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]
~~~

anijain2305 committed:
[dynamo][fallback] Fallback when run_node fails with fake tensors

[ghstack-poisoned]

ezyang commented:
Can we just fix the metas instead? For example, having the meta raise an appropriate exception type that indicates that it is definitely OK to suppress, as opposed to "oops something bad happened".

anijain2305 committed:
Update on "[dynamo][fallback] Fallback when run_node fails with fake tensors"


Little cumbersome to write a bad meta implementation for testing this. The model that I am working on falls back with this error message

Example 1
~~~
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG] Traceback (most recent call last):
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1320, in get_fake_value
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return wrap_fake_exception(
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 898, in wrap_fake_exception
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return fn()
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1321, in <lambda>
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     lambda: run_node(tx.output, node, args, kwargs, nnmodule)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1392, in run_node
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     raise RuntimeError(fn_str + str(e)).with_traceback(e.__traceback__) from e
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1379, in run_node
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return node.target(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 681, in __call__
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/utils/_stats.py", line 20, in wrapper
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return fn(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_subclasses/fake_tensor.py", line 1233, in __torch_dispatch__
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self.dispatch(func, types, args, kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_subclasses/fake_tensor.py", line 1523, in dispatch
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     r = func(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 435, in __call__
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG] RuntimeError: Failed running call_function fbgemm.dense_to_jagged(*(FakeTensor(..., device='cuda:0', size=(10, 1023, 1), dtype=torch.int64), [FakeTensor(..., device='cuda:0', size=(11,), dtype=torch.int64)]), **{}):
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG] total_L is required for meta backend
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] Graph break: Unable to propagate fake values through the operator. One possible reason could be bad meta kernel implementation. from user code at:
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/hammer/modules/features/postproc.py", line 554, in <resume in forward>
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     ) = self._action_based_sampling_impl(jt_dict)
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/nn/modules/module.py", line 1527, in _call_impl
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     return forward_call(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/hammer/modules/features/postproc.py", line 298, in forward
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     removal_mask_union_jagged = torch.ops.fbgemm.dense_to_jagged(
~~~


Example 2
~~~
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] ========== TorchDynamo Stack Trace ==========
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Traceback (most recent call last):
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1379, in run_node
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return node.target(*args, **kwargs)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 681, in __call__
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] RuntimeError: fbgemm::jagged_1d_to_dense() Expected a value of type 'int' for argument 'max_sequence_length' but instead found type 'FakeTensor'.
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Position: 2
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Value: FakeTensor(..., device='cuda:0', size=(), dtype=torch.int64)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Declaration: fbgemm::jagged_1d_to_dense(Tensor values, Tensor offsets, int max_sequence_length, int padding_value) -> Tensor
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Cast error details: Unable to cast Python instance to C++ type (compile in debug mode for details)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] The above exception was the direct cause of the following exception:
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Traceback (most recent call last):
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1320, in get_fake_value
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return wrap_fake_exception(
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 898, in wrap_fake_exception
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return fn()
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1321, in <lambda>
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     lambda: run_node(tx.output, node, args, kwargs, nnmodule)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1392, in run_node
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     raise RuntimeError(fn_str + str(e)).with_traceback(e.__traceback__) from e
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1379, in run_node
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return node.target(*args, **kwargs)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 681, in __call__
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] RuntimeError: Failed running call_function fbgemm.jagged_1d_to_dense(*(), **{'values': FakeTensor(..., device='cuda:0', size=(5110,), dtype=torch.int64), 'offsets': FakeTensor(..., device='cuda:0', size=(11,), dtype=torch.int64), 'max_sequence_length': FakeTensor(..., device='cuda:0', size=(), dtype=torch.int64), 'padding_value': 0}):
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] fbgemm::jagged_1d_to_dense() Expected a value of type 'int' for argument 'max_sequence_length' but instead found type 'FakeTensor'.
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Position: 2
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Value: FakeTensor(..., device='cuda:0', size=(), dtype=torch.int64)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Declaration: fbgemm::jagged_1d_to_dense(Tensor values, Tensor offsets, int max_sequence_length, int padding_value) -> Tensor
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Cast error details: Unable to cast Python instance to C++ type (compile in debug mode for details)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] Graph break: Unable to propagate fake values through the operator. One possible reason could be bad meta kernel implementation. from user code at:
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/hammer/modules/features/postproc.py", line 141, in <resume in flip_batched_id_list_feature>
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     dense_values = torch.ops.fbgemm.jagged_1d_to_dense(
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]
~~~


[ghstack-poisoned]

anijain2305 commented:
> Can we just fix the metas instead? For example, having the meta raise an appropriate exception type that indicates that it is definitely OK to suppress, as opposed to "oops something bad happened".

The errors that I am seeing are coming from C++. I am not sure how to catch these and convert them into Python "allowed" exceptions. So, maybe these are not good candidates for fallback?

anijain2305 committed:
Update on "[dynamo][fallback] Fallback when run_node fails with fake tensors"


Little cumbersome to write a bad meta implementation for testing this. The model that I am working on falls back with this error message

Example 1
~~~
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG] Traceback (most recent call last):
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1320, in get_fake_value
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return wrap_fake_exception(
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 898, in wrap_fake_exception
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return fn()
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1321, in <lambda>
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     lambda: run_node(tx.output, node, args, kwargs, nnmodule)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1392, in run_node
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     raise RuntimeError(fn_str + str(e)).with_traceback(e.__traceback__) from e
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1379, in run_node
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return node.target(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 681, in __call__
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/utils/_stats.py", line 20, in wrapper
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return fn(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_subclasses/fake_tensor.py", line 1233, in __torch_dispatch__
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self.dispatch(func, types, args, kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_subclasses/fake_tensor.py", line 1523, in dispatch
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     r = func(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 435, in __call__
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG] RuntimeError: Failed running call_function fbgemm.dense_to_jagged(*(FakeTensor(..., device='cuda:0', size=(10, 1023, 1), dtype=torch.int64), [FakeTensor(..., device='cuda:0', size=(11,), dtype=torch.int64)]), **{}):
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG] total_L is required for meta backend
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] Graph break: Unable to propagate fake values through the operator. One possible reason could be bad meta kernel implementation. from user code at:
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/hammer/modules/features/postproc.py", line 554, in <resume in forward>
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     ) = self._action_based_sampling_impl(jt_dict)
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/nn/modules/module.py", line 1527, in _call_impl
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     return forward_call(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/hammer/modules/features/postproc.py", line 298, in forward
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     removal_mask_union_jagged = torch.ops.fbgemm.dense_to_jagged(
~~~


Example 2
~~~
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] ========== TorchDynamo Stack Trace ==========
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Traceback (most recent call last):
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1379, in run_node
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return node.target(*args, **kwargs)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 681, in __call__
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] RuntimeError: fbgemm::jagged_1d_to_dense() Expected a value of type 'int' for argument 'max_sequence_length' but instead found type 'FakeTensor'.
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Position: 2
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Value: FakeTensor(..., device='cuda:0', size=(), dtype=torch.int64)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Declaration: fbgemm::jagged_1d_to_dense(Tensor values, Tensor offsets, int max_sequence_length, int padding_value) -> Tensor
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Cast error details: Unable to cast Python instance to C++ type (compile in debug mode for details)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] The above exception was the direct cause of the following exception:
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Traceback (most recent call last):
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1320, in get_fake_value
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return wrap_fake_exception(
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 898, in wrap_fake_exception
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return fn()
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1321, in <lambda>
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     lambda: run_node(tx.output, node, args, kwargs, nnmodule)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1392, in run_node
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     raise RuntimeError(fn_str + str(e)).with_traceback(e.__traceback__) from e
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1379, in run_node
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return node.target(*args, **kwargs)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 681, in __call__
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] RuntimeError: Failed running call_function fbgemm.jagged_1d_to_dense(*(), **{'values': FakeTensor(..., device='cuda:0', size=(5110,), dtype=torch.int64), 'offsets': FakeTensor(..., device='cuda:0', size=(11,), dtype=torch.int64), 'max_sequence_length': FakeTensor(..., device='cuda:0', size=(), dtype=torch.int64), 'padding_value': 0}):
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] fbgemm::jagged_1d_to_dense() Expected a value of type 'int' for argument 'max_sequence_length' but instead found type 'FakeTensor'.
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Position: 2
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Value: FakeTensor(..., device='cuda:0', size=(), dtype=torch.int64)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Declaration: fbgemm::jagged_1d_to_dense(Tensor values, Tensor offsets, int max_sequence_length, int padding_value) -> Tensor
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Cast error details: Unable to cast Python instance to C++ type (compile in debug mode for details)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] Graph break: Unable to propagate fake values through the operator. One possible reason could be bad meta kernel implementation. from user code at:
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/hammer/modules/features/postproc.py", line 141, in <resume in flip_batched_id_list_feature>
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     dense_values = torch.ops.fbgemm.jagged_1d_to_dense(
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]
~~~


[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][fallback] Fallback when run_node fails with fake tensors"


Little cumbersome to write a bad meta implementation for testing this. The model that I am working on falls back with this error message

Example 1
~~~
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG] Traceback (most recent call last):
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1320, in get_fake_value
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return wrap_fake_exception(
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 898, in wrap_fake_exception
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return fn()
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1321, in <lambda>
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     lambda: run_node(tx.output, node, args, kwargs, nnmodule)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1392, in run_node
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     raise RuntimeError(fn_str + str(e)).with_traceback(e.__traceback__) from e
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1379, in run_node
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return node.target(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 681, in __call__
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/utils/_stats.py", line 20, in wrapper
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return fn(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_subclasses/fake_tensor.py", line 1233, in __torch_dispatch__
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self.dispatch(func, types, args, kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_subclasses/fake_tensor.py", line 1523, in dispatch
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     r = func(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 435, in __call__
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG] RuntimeError: Failed running call_function fbgemm.dense_to_jagged(*(FakeTensor(..., device='cuda:0', size=(10, 1023, 1), dtype=torch.int64), [FakeTensor(..., device='cuda:0', size=(11,), dtype=torch.int64)]), **{}):
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG] total_L is required for meta backend
[rank0]:[2023-08-14 22:54:39,330] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] Graph break: Unable to propagate fake values through the operator. One possible reason could be bad meta kernel implementation. from user code at:
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/hammer/modules/features/postproc.py", line 554, in <resume in forward>
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     ) = self._action_based_sampling_impl(jt_dict)
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/nn/modules/module.py", line 1527, in _call_impl
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     return forward_call(*args, **kwargs)
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/hammer/modules/features/postproc.py", line 298, in forward
[rank0]:[2023-08-14 22:54:39,331] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     removal_mask_union_jagged = torch.ops.fbgemm.dense_to_jagged(
~~~


Example 2
~~~
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] ========== TorchDynamo Stack Trace ==========
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Traceback (most recent call last):
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1379, in run_node
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return node.target(*args, **kwargs)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 681, in __call__
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] RuntimeError: fbgemm::jagged_1d_to_dense() Expected a value of type 'int' for argument 'max_sequence_length' but instead found type 'FakeTensor'.
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Position: 2
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Value: FakeTensor(..., device='cuda:0', size=(), dtype=torch.int64)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Declaration: fbgemm::jagged_1d_to_dense(Tensor values, Tensor offsets, int max_sequence_length, int padding_value) -> Tensor
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Cast error details: Unable to cast Python instance to C++ type (compile in debug mode for details)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] The above exception was the direct cause of the following exception:
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Traceback (most recent call last):
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1320, in get_fake_value
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return wrap_fake_exception(
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 898, in wrap_fake_exception
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return fn()
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1321, in <lambda>
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     lambda: run_node(tx.output, node, args, kwargs, nnmodule)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1392, in run_node
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     raise RuntimeError(fn_str + str(e)).with_traceback(e.__traceback__) from e
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_dynamo/utils.py", line 1379, in run_node
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return node.target(*args, **kwargs)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/torch/_ops.py", line 681, in __call__
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]     return self._op(*args, **kwargs or {})
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] RuntimeError: Failed running call_function fbgemm.jagged_1d_to_dense(*(), **{'values': FakeTensor(..., device='cuda:0', size=(5110,), dtype=torch.int64), 'offsets': FakeTensor(..., device='cuda:0', size=(11,), dtype=torch.int64), 'max_sequence_length': FakeTensor(..., device='cuda:0', size=(), dtype=torch.int64), 'padding_value': 0}):
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] fbgemm::jagged_1d_to_dense() Expected a value of type 'int' for argument 'max_sequence_length' but instead found type 'FakeTensor'.
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Position: 2
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Value: FakeTensor(..., device='cuda:0', size=(), dtype=torch.int64)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Declaration: fbgemm::jagged_1d_to_dense(Tensor values, Tensor offsets, int max_sequence_length, int padding_value) -> Tensor
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG] Cast error details: Unable to cast Python instance to C++ type (compile in debug mode for details)
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.exc.__graph_breaks: [DEBUG]
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] Graph break: Unable to propagate fake values through the operator. One possible reason could be bad meta kernel implementation. from user code at:
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]   File "/mnt/xarfuse/uid-416112/567437c7-seed-nspid4026531836_cgpid49144894-ns-4026531840/hammer/modules/features/postproc.py", line 141, in <resume in flip_batched_id_list_feature>
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     dense_values = torch.ops.fbgemm.jagged_1d_to_dense(
[rank0]:[2023-08-14 22:54:43,443] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]
~~~


[ghstack-poisoned]

ezyang commented:
> The errors that I am seeing are also coming from C++. I am not sure how to catch those and convert them into Python "allowed" exceptions. So, maybe these are not good candidates for fallback?

We have variants of the TORCH_CHECK macros which can raise alternate exception types. In your case `TORCH_CHECK_NOT_IMPLEMENTED` would be pretty appropriate (and then you just have to catch NotImplementedError only)
